src/main/scala/viper/gobra/ast/internal/Program.scala

```diff
@@ -267,9 +267,24 @@ case class Seqn(stmts: Vector[Stmt])(val info: Source.Parser.Info) extends Stmt
 
 case class Label(id: LabelProxy)(val info: Source.Parser.Info) extends Stmt
 
+/**
+  * 'label' corresponds to the loop label we want to continue. In case it is a normal
+  * continue node it is None.
+  * The continue node will be replaced by a goto statement to 'escLabel' which has
+  * been placed properly while desugaring for loops.
+  */
+case class Continue(label: Option[String], escLabel: String)(val info: Source.Parser.Info) extends Stmt
+
+/**
+  * 'label' corresponds to the loop label we want to break out of. In case it is a normal
+  * break node it is 'None'.
+  * The break node will be replaced by a goto statement to 'escLabel'.
+  */
+case class Break(label: Option[String], escLabel: String)(val info: Source.Parser.Info) extends Stmt
```

Comment:
The current approach looks a bit too complicated to me. I would just transform a loop label into three labels:

```
L:
<loop>
```
to
```
L:
L$Continue:
<loop>
L$Break
```

For loops that do not have a label you just take a fresh identifier `F` and create the `F$Continue` and `F$Break` label. Similarly, for unlabeled break and continues, you just add the corresponding fresh identifier and treat them as if they had a label.

And then tranform 
`goto L` to `goto L`
`break L` to `goto L$Break`
`continue L` to `goto L$Continue`

You can do that either in the desugaring (then you only need a goto node in the internal AST and loops also do not have a label) or in the encoding (then the internal AST has goto, break, continue, but they have exactly one label, and loops have a label). I prefer the solution within the desugarer. Also you should always use `LabelProxy` instead of `String`. Laslty, it would be interesting if too many labels cause performance problems. But we could address this with a Viper to Viper transformation, which eliminates unused labels or maybe even merges labels together if possible.

Reply:
I just noticed that my continue label is wrong, but you get the idea (I guess you put it at the end of the loop body)